File: packages/saju/src/core/strength.ts
import type { Label } from "@/types";
import {
  type Element,
  getStemElement,
  getStemPolarity,
  getTenGodKey,
  type TenGodKey,
} from "./ten-gods";

export const STRENGTH_LEVEL_KEYS = [
  "extremelyWeak",
  "veryWeak",
  "weak",
  "neutralWeak",
  "neutral",
  "neutralStrong",
  "strong",
  "veryStrong",
  "extremelyStrong",
] as const;

export type StrengthLevelKey = (typeof STRENGTH_LEVEL_KEYS)[number];

export interface StrengthLevelLabel extends Label<StrengthLevelKey> {}

const STRENGTH_LEVEL_DATA: Record<StrengthLevelKey, { korean: string; hanja: string }> = {
  extremelyWeak: { korean: "극약", hanja: "極弱" },
  veryWeak: { korean: "태약", hanja: "太弱" },
  weak: { korean: "신약", hanja: "身弱" },
  neutralWeak: { korean: "중화신약", hanja: "中和身弱" },
  neutral: { korean: "중화", hanja: "中和" },
  neutralStrong: { korean: "중화신강", hanja: "中和身強" },
  strong: { korean: "신강", hanja: "身強" },
  veryStrong: { korean: "태강", hanja: "太強" },
  extremelyStrong: { korean: "극왕", hanja: "極旺" },
};

export function getStrengthLevelLabel(key: StrengthLevelKey): StrengthLevelLabel {
  const data = STRENGTH_LEVEL_DATA[key];
  return { key, ...data };
}

interface HiddenStemWeight {
  stem: string;
  weight: number;
  type: "본기" | "중기" | "여기";
}

const HIDDEN_STEM_WEIGHTS: Record<string, HiddenStemWeight[]> = {
  子: [{ stem: "癸", weight: 1.0, type: "본기" }],
  丑: [
    { stem: "己", weight: 0.6, type: "본기" },
    { stem: "癸", weight: 0.25, type: "중기" },
    { stem: "辛", weight: 0.15, type: "여기" },
  ],
  寅: [
    { stem: "甲", weight: 0.6, type: "본기" },
    { stem: "丙", weight: 0.25, type: "중기" },
    { stem: "戊", weight: 0.15, type: "여기" },
  ],
  卯: [{ stem: "乙", weight: 1.0, type: "본기" }],
  辰: [
    { stem: "戊", weight: 0.6, type: "본기" },
    { stem: "乙", weight: 0.25, type: "중기" },
    { stem: "癸", weight: 0.15, type: "여기" },
  ],
  巳: [
    { stem: "丙", weight: 0.6, type: "본기" },
    { stem: "庚", weight: 0.25, type: "중기" },
    { stem: "戊", weight: 0.15, type: "여기" },
  ],
  午: [
    { stem: "丁", weight: 0.7, type: "본기" },
    { stem: "己", weight: 0.3, type: "중기" },
  ],
  未: [
    { stem: "己", weight: 0.6, type: "본기" },
    { stem: "丁", weight: 0.25, type: "중기" },
    { stem: "乙", weight: 0.15, type: "여기" },
  ],
  申: [
    { stem: "庚", weight: 0.6, type: "본기" },
    { stem: "壬", weight: 0.25, type: "중기" },
    { stem: "戊", weight: 0.15, type: "여기" },
  ],
  酉: [{ stem: "辛", weight: 1.0, type: "본기" }],
  戌: [
    { stem: "戊", weight: 0.6, type: "본기" },
    { stem: "辛", weight: 0.25, type: "중기" },
    { stem: "丁", weight: 0.15, type: "여기" },
  ],
  亥: [
    { stem: "壬", weight: 0.7, type: "본기" },
    { stem: "甲", weight: 0.3, type: "중기" },
  ],
};

type SeasonalElement = "wood" | "fire" | "wet_earth" | "dry_earth" | "metal" | "water";

const MONTH_BRANCH_SEASONAL: Record<string, SeasonalElement> = {
  寅: "wood",
  卯: "wood",
  辰: "wet_earth",
  巳: "fire",
  午: "fire",
  未: "dry_earth",
  申: "metal",
  酉: "metal",
  戌: "dry_earth",
  亥: "water",
  子: "water",
  丑: "wet_earth",
};

function getMonthlyStrengthMultiplier(dayMasterElement: Element, monthBranch: string): number {
  const seasonal = MONTH_BRANCH_SEASONAL[monthBranch];

  const strengthTable: Record<Element, Record<SeasonalElement, number>> = {
    wood: {
      wood: 1.0,
      fire: 0.3,
      wet_earth: 0.5,
      dry_earth: 0.2,
      metal: 0.1,
      water: 0.7,
    },
    fire: {
      wood: 0.7,
      fire: 1.0,
      wet_earth: 0.3,
      dry_earth: 0.5,
      metal: 0.1,
      water: 0.1,
    },
    earth: {
      wood: 0.1,
      fire: 0.7,
      wet_earth: 0.8,
      dry_earth: 1.0,
      metal: 0.3,
      water: 0.1,
    },
    metal: {
      wood: 0.1,
      fire: 0.1,
      wet_earth: 0.5,
      dry_earth: 0.7,
      metal: 1.0,
      water: 0.3,
    },
    water: {
      wood: 0.3,
      fire: 0.1,
      wet_earth: 0.2,
      dry_earth: 0.1,
      metal: 0.7,
      water: 1.0,
    },
  };

  return strengthTable[dayMasterElement][seasonal];
}

function calculateRootStrength(dayMaster: string, branch: string): number {
  const dayMasterElement = getStemElement(dayMaster);
  const dayMasterPolarity = getStemPolarity(dayMaster);

  const hiddenStems = HIDDEN_STEM_WEIGHTS[branch] || [];
  let strength = 0;

  for (const hs of hiddenStems) {
    const hsElement = getStemElement(hs.stem);
    const hsPolarity = getStemPolarity(hs.stem);

    if (hsElement === dayMasterElement) {
      if (hsPolarity === dayMasterPolarity) {
        strength += hs.weight * 1.0;
      } else {
        strength += hs.weight * 0.7;
      }
    } else if (
      (dayMasterElement === "wood" && hsElement === "water") ||
      (dayMasterElement === "fire" && hsElement === "wood") ||
      (dayMasterElement === "earth" && hsElement === "fire") ||
      (dayMasterElement === "metal" && hsElement === "earth") ||
      (dayMasterElement === "water" && hsElement === "metal")
    ) {
      strength += hs.weight * 0.5;
    }
  }

  return strength;
}

function isTransparent(stem: string, allStems: string[]): boolean {
  return allStems.includes(stem);
}

const HELPFUL_TEN_GODS: TenGodKey[] = ["companion", "robWealth", "directSeal", "indirectSeal"];
const WEAKENING_TEN_GODS: TenGodKey[] = [
  "eatingGod",
  "hurtingOfficer",
  "indirectWealth",
  "directWealth",
  "sevenKillings",
  "directOfficer",
];

function isHelpfulTenGod(tenGod: TenGodKey): boolean {
  return HELPFUL_TEN_GODS.includes(tenGod);
}

function isWeakeningTenGod(tenGod: TenGodKey): boolean {
  return WEAKENING_TEN_GODS.includes(tenGod);
}

export interface StrengthFactors {
  deukryeong: number;
  deukji: number;
  deukse: number;
  tonggeun: number;
  helpCount: number;
  weakenCount: number;
}

export interface StrengthResult {
  level: StrengthLevelLabel;
  score: number;
  factors: StrengthFactors;
  description: string;
}

export function analyzeStrength(
  yearPillar: string,
  monthPillar: string,
  dayPillar: string,
  hourPillar: string,
): StrengthResult {
  const dayMaster = dayPillar[0];
  const dayMasterElement = getStemElement(dayMaster);
  const monthBranch = monthPillar[1];

  const deukryeong = getMonthlyStrengthMultiplier(dayMasterElement, monthBranch);

  const allBranches = [yearPillar[1], monthPillar[1], dayPillar[1], hourPillar[1]];
  let tonggeun = 0;
  for (const branch of allBranches) {
    tonggeun += calculateRootStrength(dayMaster, branch);
  }

  const allStems = [yearPillar[0], monthPillar[0], dayPillar[0], hourPillar[0]];

  const monthHiddenStems = HIDDEN_STEM_WEIGHTS[monthBranch] || [];
  let transparentBonus = 0;
  for (const hs of monthHiddenStems) {
    if (isTransparent(hs.stem, allStems)) {
      const tenGod = getTenGodKey(dayMaster, hs.stem);
      if (isHelpfulTenGod(tenGod)) {
        transparentBonus += hs.weight * 0.3;
      }
    }
  }

  let deukji = 0;
  const nearBranches = [dayPillar[1], hourPillar[1]];
  for (const branch of nearBranches) {
    deukji += calculateRootStrength(dayMaster, branch);
  }

  let deukse = 0;
  const otherStems = [yearPillar[0], monthPillar[0], hourPillar[0]];
  for (const stem of otherStems) {
    const tenGod = getTenGodKey(dayMaster, stem);
    if (isHelpfulTenGod(tenGod)) {
      deukse += 1;
    }
  }

  let helpCount = 0;
  let weakenCount = 0;

  for (const stem of otherStems) {
    const tenGod = getTenGodKey(dayMaster, stem);
    if (isHelpfulTenGod(tenGod)) helpCount++;
    if (isWeakeningTenGod(tenGod)) weakenCount++;
  }

  for (const branch of allBranches) {
    const hiddenStems = HIDDEN_STEM_WEIGHTS[branch] || [];
    const mainStem = hiddenStems[0]?.stem;
    if (mainStem) {
      const tenGod = getTenGodKey(dayMaster, mainStem);
      if (isHelpfulTenGod(tenGod)) helpCount++;
      if (isWeakeningTenGod(tenGod)) weakenCount++;
    }
  }

  let score = 0;

  score += deukryeong * 35;
  score += tonggeun * 20;
  score += transparentBonus * 15;
  score += deukse * 8;
  score += helpCount * 5;
  score -= weakenCount * 6;

  score = Math.round(score * 10) / 10;

  let levelKey: StrengthLevelKey;
  if (score <= 10) levelKey = "extremelyWeak";
  else if (score <= 20) levelKey = "veryWeak";
  else if (score <= 30) levelKey = "weak";
  else if (score <= 38) levelKey = "neutralWeak";
  else if (score <= 45) levelKey = "neutral";
  else if (score <= 55) levelKey = "neutralStrong";
  else if (score <= 70) levelKey = "strong";
  else if (score <= 85) levelKey = "veryStrong";
  else levelKey = "extremelyStrong";

  const factors: StrengthFactors = {
    deukryeong: Math.round(deukryeong * 100) / 100,
    deukji: Math.round(deukji * 100) / 100,
    deukse,
    tonggeun: Math.round(tonggeun * 100) / 100,
    helpCount,
    weakenCount,
  };

  const deukryeongDesc = deukryeong >= 0.7 ? "득령" : deukryeong >= 0.4 ? "반득령" : "실령";
  let description = `일간 ${dayMaster}(${dayMasterElement}), ${deukryeongDesc}(${Math.round(deukryeong * 100)}%)`;
  description += tonggeun > 0 ? `, 통근(${Math.round(tonggeun * 100) / 100})` : "";
  description += deukse > 0 ? `, 득세(${deukse})` : "";

  return {
    level: getStrengthLevelLabel(levelKey),
    score,
    factors,
    description,
  };
}

File: packages/saju-dart/lib/src/core/strength.dart
import '../types/types.dart';
import 'ten_gods.dart';

/// Seasonal element type (includes wet/dry earth distinction)
enum _SeasonalElement {
  wood,
  fire,
  wetEarth,
  dryEarth,
  metal,
  water,
}

/// Get seasonal element from month branch
_SeasonalElement _getSeasonalElement(Branch monthBranch) {
  return switch (monthBranch) {
    Branch.yin => _SeasonalElement.wood,
    Branch.mao => _SeasonalElement.wood,
    Branch.chen => _SeasonalElement.wetEarth,
    Branch.si => _SeasonalElement.fire,
    Branch.wu => _SeasonalElement.fire,
    Branch.wei => _SeasonalElement.dryEarth,
    Branch.shen => _SeasonalElement.metal,
    Branch.you => _SeasonalElement.metal,
    Branch.xu => _SeasonalElement.dryEarth,
    Branch.hai => _SeasonalElement.water,
    Branch.zi => _SeasonalElement.water,
    Branch.chou => _SeasonalElement.wetEarth,
  };
}

/// Monthly strength multiplier table
double _getMonthlyStrengthMultiplier(Element dayMasterElement, Branch monthBranch) {
  final seasonal = _getSeasonalElement(monthBranch);

  final strengthTable = <Element, Map<_SeasonalElement, double>>{
    Element.wood: {
      _SeasonalElement.wood: 1.0,
      _SeasonalElement.fire: 0.3,
      _SeasonalElement.wetEarth: 0.5,
      _SeasonalElement.dryEarth: 0.2,
      _SeasonalElement.metal: 0.1,
      _SeasonalElement.water: 0.7,
    },
    Element.fire: {
      _SeasonalElement.wood: 0.7,
      _SeasonalElement.fire: 1.0,
      _SeasonalElement.wetEarth: 0.3,
      _SeasonalElement.dryEarth: 0.5,
      _SeasonalElement.metal: 0.1,
      _SeasonalElement.water: 0.1,
    },
    Element.earth: {
      _SeasonalElement.wood: 0.1,
      _SeasonalElement.fire: 0.7,
      _SeasonalElement.wetEarth: 0.8,
      _SeasonalElement.dryEarth: 1.0,
      _SeasonalElement.metal: 0.3,
      _SeasonalElement.water: 0.1,
    },
    Element.metal: {
      _SeasonalElement.wood: 0.1,
      _SeasonalElement.fire: 0.1,
      _SeasonalElement.wetEarth: 0.5,
      _SeasonalElement.dryEarth: 0.7,
      _SeasonalElement.metal: 1.0,
      _SeasonalElement.water: 0.3,
    },
    Element.water: {
      _SeasonalElement.wood: 0.3,
      _SeasonalElement.fire: 0.1,
      _SeasonalElement.wetEarth: 0.2,
      _SeasonalElement.dryEarth: 0.1,
      _SeasonalElement.metal: 0.7,
      _SeasonalElement.water: 1.0,
    },
  };

  return strengthTable[dayMasterElement]?[seasonal] ?? 0.5;
}

/// Calculate root strength (통근) for a branch
double _calculateRootStrength(Stem dayMaster, Branch branch) {
  final dayMasterElement = dayMaster.element;
  final dayMasterPolarity = dayMaster.polarity;

  final hiddenStems = HiddenStems.forBranch(branch);
  var strength = 0.0;

  for (final hs in hiddenStems) {
    final hsElement = hs.stem.element;
    final hsPolarity = hs.stem.polarity;

    if (hsElement == dayMasterElement) {
      // Same element
      if (hsPolarity == dayMasterPolarity) {
        strength += hs.weight * 1.0;
      } else {
        strength += hs.weight * 0.7;
      }
    } else if (hsElement.generates == dayMasterElement) {
      // Hidden stem generates day master
      strength += hs.weight * 0.5;
    }
  }

  return strength;
}

/// Check if stem appears in the list
bool _isTransparent(Stem stem, List<Stem> allStems) {
  return allStems.contains(stem);
}

/// Helpful ten gods (help day master)
const _helpfulTenGods = [
  TenGod.companion,
  TenGod.robWealth,
  TenGod.directSeal,
  TenGod.indirectSeal,
];

/// Weakening ten gods
const _weakeningTenGods = [
  TenGod.eatingGod,
  TenGod.hurtingOfficer,
  TenGod.indirectWealth,
  TenGod.directWealth,
  TenGod.sevenKillings,
  TenGod.directOfficer,
];

bool _isHelpfulTenGod(TenGod tenGod) => _helpfulTenGods.contains(tenGod);
bool _isWeakeningTenGod(TenGod tenGod) => _weakeningTenGods.contains(tenGod);

/// Strength factors
class StrengthFactors {
  const StrengthFactors({
    required this.deukryeong,
    required this.deukji,
    required this.deukse,
    required this.tonggeun,
    required this.helpCount,
    required this.weakenCount,
  });

  /// Monthly strength (득령) 0-1
  final double deukryeong;

  /// Position strength (득지)
  final double deukji;

  /// Supporting stems (득세)
  final int deukse;

  /// Root strength (통근)
  final double tonggeun;

  /// Helpful ten gods count
  final int helpCount;

  /// Weakening ten gods count
  final int weakenCount;
}

/// Strength analysis result
class StrengthResult {
  const StrengthResult({
    required this.level,
    required this.score,
    required this.factors,
    required this.description,
  });

  final StrengthLevel level;
  final double score;
  final StrengthFactors factors;
  final String description;
}

/// Analyze day master strength
StrengthResult analyzeStrength(FourPillars pillars) {
  final dayMaster = pillars.dayMaster;
  final dayMasterElement = dayMaster.element;
  final monthBranch = pillars.month.branch;

  // 1. 득령 (deukryeong) - monthly strength
  final deukryeong = _getMonthlyStrengthMultiplier(dayMasterElement, monthBranch);

  // 2. 통근 (tonggeun) - root strength in all branches
  var tonggeun = 0.0;
  for (final branch in pillars.branches) {
    tonggeun += _calculateRootStrength(dayMaster, branch);
  }

  // 3. 투간 (transparent bonus) - hidden stems appearing in pillars
  final allStems = pillars.stems;
  final monthHiddenStems = HiddenStems.forBranch(monthBranch);
  var transparentBonus = 0.0;
  for (final hs in monthHiddenStems) {
    if (_isTransparent(hs.stem, allStems)) {
      final tenGod = getTenGodKey(dayMaster, hs.stem);
      if (_isHelpfulTenGod(tenGod)) {
        transparentBonus += hs.weight * 0.3;
      }
    }
  }

  // 4. 득지 (deukji) - position strength (day and hour branches)
  var deukji = 0.0;
  deukji += _calculateRootStrength(dayMaster, pillars.day.branch);
  deukji += _calculateRootStrength(dayMaster, pillars.hour.branch);

  // 5. 득세 (deukse) - supporting stems count
  var deukse = 0;
  final otherStems = [pillars.year.stem, pillars.month.stem, pillars.hour.stem];
  for (final stem in otherStems) {
    final tenGod = getTenGodKey(dayMaster, stem);
    if (_isHelpfulTenGod(tenGod)) {
      deukse += 1;
    }
  }

  // 6. Count helpful and weakening ten gods
  var helpCount = 0;
  var weakenCount = 0;

  for (final stem in otherStems) {
    final tenGod = getTenGodKey(dayMaster, stem);
    if (_isHelpfulTenGod(tenGod)) helpCount++;
    if (_isWeakeningTenGod(tenGod)) weakenCount++;
  }

  for (final branch in pillars.branches) {
    final hiddenStems = HiddenStems.forBranch(branch);
    if (hiddenStems.isNotEmpty) {
      final mainStem = hiddenStems.first.stem;
      final tenGod = getTenGodKey(dayMaster, mainStem);
      if (_isHelpfulTenGod(tenGod)) helpCount++;
      if (_isWeakeningTenGod(tenGod)) weakenCount++;
    }
  }

  // Calculate score
  var score = 0.0;
  score += deukryeong * 35;
  score += tonggeun * 20;
  score += transparentBonus * 15;
  score += deukse * 8;
  score += helpCount * 5;
  score -= weakenCount * 6;
  score = (score * 10).round() / 10;

  // Determine strength level
  final level = switch (score) {
    <= 10 => StrengthLevel.extremelyWeak,
    <= 20 => StrengthLevel.veryWeak,
    <= 30 => StrengthLevel.weak,
    <= 38 => StrengthLevel.neutralWeak,
    <= 45 => StrengthLevel.neutral,
    <= 55 => StrengthLevel.neutralStrong,
    <= 70 => StrengthLevel.strong,
    <= 85 => StrengthLevel.veryStrong,
    _ => StrengthLevel.extremelyStrong,
  };

  final factors = StrengthFactors(
    deukryeong: (deukryeong * 100).round() / 100,
    deukji: (deukji * 100).round() / 100,
    deukse: deukse,
    tonggeun: (tonggeun * 100).round() / 100,
    helpCount: helpCount,
    weakenCount: weakenCount,
  );

  // Build description
  final deukryeongDesc =
      deukryeong >= 0.7 ? '득령' : (deukryeong >= 0.4 ? '반득령' : '실령');
  var description =
      '일간 ${dayMaster.hanja}(${dayMasterElement.korean}), $deukryeongDesc(${(deukryeong * 100).round()}%)';
  if (tonggeun > 0) {
    description += ', 통근(${(tonggeun * 100).round() / 100})';
  }
  if (deukse > 0) {
    description += ', 득세($deukse)';
  }

  return StrengthResult(
    level: level,
    score: score,
    factors: factors,
    description: description,
  );
}
